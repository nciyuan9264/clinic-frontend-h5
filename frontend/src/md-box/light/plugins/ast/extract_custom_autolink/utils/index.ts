export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
