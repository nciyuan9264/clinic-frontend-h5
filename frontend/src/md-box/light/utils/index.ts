export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
