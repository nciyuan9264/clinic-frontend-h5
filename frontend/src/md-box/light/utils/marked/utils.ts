/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
