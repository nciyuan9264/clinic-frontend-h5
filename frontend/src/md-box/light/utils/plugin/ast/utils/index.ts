export * from './update_ast';
