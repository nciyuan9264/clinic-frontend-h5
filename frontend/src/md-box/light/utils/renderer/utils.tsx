import React, { createElement } from 'react';

import { isRegExp } from 'lodash-es';
import { DOMNode, Element, Text, domToReact } from 'html-react-parser';

import { RenderFunction } from '../../type';

export const isTextType = (node: DOMNode): node is Text => {
  return node.type === 'text';
};

export const isEndlineText = (node: DOMNode): node is Text => {
  return isTextType(node) && /^\n+$/.test(getInnerText(node));
};

/**
 * 此处断言的是Element的子类型：即为A标签的类型，但是Element不可进一步进行类型收窄，因此
 * 返回node is Element
 */
export const isElementType = (
  node: DOMNode,
  ...names: (string | RegExp)[]
): node is Element => {
  if (!('attribs' in node) || typeof node.attribs !== 'object') {
    return false;
  }
  return (
    node.type === 'tag' &&
    (!names?.length ||
      names.some((item) => {
        if (isRegExp(item)) {
          return item.test(node.name);
        } else {
          return item === node.name;
        }
      }))
  );
};

export const getInnerText = (node: DOMNode): string => {
  if (isTextType(node)) {
    return node.data;
  }
  if (isElementType(node)) {
    return node.children.map(getInnerText).join('');
  }
  return '';
};

/**
 * 检测字符或字符串是否包含中文字符
 * https://github.com/anton-bot/contains-chinese/blob/master/index.js#L7
 * @param char 字符或字符串
 * @returns 是否包含中文字符
 */
export const detectIsChinese = (char: string) => {
  const HAN_REGEX =
    // eslint-disable-next-line max-len
    /[\u2E80-\u2E99\u2E9B-\u2EF3\u2F00-\u2FD5\u3005\u3007\u3021-\u3029\u3038-\u303B\u3400-\u4DB5\u4E00-\u9FD5\uF900-\uFA6D\uFA70-\uFAD9]/;
  return HAN_REGEX.test(char);
};

export const renderDom: RenderFunction<false> = (node) => {
  return <>{domToReact([node])}</>;
};

export const renderReactElement = <P extends Record<string, unknown>>(
  tag: string,
  props: P,
  children?: React.ReactNode[],
) => {
  if (props.className === 'markdown-root') {
    return <>{children}</>;
  } else {
    return (
      <>{createElement(tag, props, children?.length ? children : undefined)}</>
    );
  }
};
