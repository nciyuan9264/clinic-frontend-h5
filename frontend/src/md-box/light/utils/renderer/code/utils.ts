import { DOMNode, Element } from 'html-react-parser';

import { isElementType } from '../utils';

const DEFAULT_LANGUAGE = 'plaintext';

const getCodeElement = (node: DOMNode): Element | undefined => {
  if (isElementType(node, 'code')) {
    return node;
  }

  if (isElementType(node, 'pre')) {
    return node.childNodes.find((childNode): childNode is Element =>
      isElementType(childNode, 'code'),
    );
  }
};

export const getLanguageFromElement = (
  node: DOMNode,
  defaultLanguage = DEFAULT_LANGUAGE,
): string => {
  const codeElement = getCodeElement(node);

  if (codeElement) {
    return (
      codeElement.attribs?.class?.match(/language-(.*)/)?.[1] || defaultLanguage
    );
  }

  return defaultLanguage;
};

export const getMetaFromElement = (node: DOMNode, removeBuiltIn = true) => {
  const codeElement = getCodeElement(node);

  const metaString = codeElement?.attribs['data-meta'];

  if (!metaString) {
    return;
  }

  if (!removeBuiltIn) {
    return metaString;
  }

  return (
    metaString
      .split(/\s+/)
      .filter((item) => !item.match(/^__[^_].*/))
      .join(' ') || undefined
  );
};

export const getIndentFromCodeElememt = (node: DOMNode) => {
  const metaString = getMetaFromElement(node, false);

  if (!metaString) {
    return 0;
  }

  const matchResult = metaString.match(
    /(\s|^)__indent=(?<indent>[0-9]+)(\s|$)/,
  );

  if (!matchResult) {
    return 0;
  }

  const indentValue = Number(matchResult.groups?.indent);

  return isNaN(indentValue) ? 0 : indentValue;
};
