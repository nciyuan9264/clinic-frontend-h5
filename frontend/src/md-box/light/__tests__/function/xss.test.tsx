import { render, act } from '@testing-library/react';

import { MdBoxForTesting } from '../utils';

describe('防止 XSS 注入测试', () => {
  test('注入 latex href 命令，验证注入不成功', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={'$$ href{javascript:alert(1)}{Click Me!} $$'}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });
});
