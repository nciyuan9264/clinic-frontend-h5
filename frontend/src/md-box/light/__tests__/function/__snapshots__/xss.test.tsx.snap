// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`防止 XSS 注入测试 注入 latex href 命令，验证注入不成功 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <span
        class="container math-inline"
        data-custom-copy-text="\\(href{javascript:alert(1)}{Click Me!}\\)"
      >
        <svg
          height="2.262ex"
          role="img"
          style="vertical-align: -0.566ex;"
          viewBox="0 -750 14928.6 1000"
          width="33.775ex"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="currentColor"
            stroke="currentColor"
            stroke-width="0"
            transform="scale(1,-1)"
          >
            <g
              data-mml-node="math"
            >
              <g
                data-mml-node="mi"
              >
                <path
                  d="M137 683Q138 683 209 688T282 694Q294 694 294 685Q294 674 258 534Q220 386 220 383Q220 381 227 388Q288 442 357 442Q411 442 444 415T478 336Q478 285 440 178T402 50Q403 36 407 31T422 26Q450 26 474 56T513 138Q516 149 519 151T535 153Q555 153 555 145Q555 144 551 130Q535 71 500 33Q466 -10 419 -10H414Q367 -10 346 17T325 74Q325 90 361 192T398 345Q398 404 354 404H349Q266 404 205 306L198 293L164 158Q132 28 127 16Q114 -11 83 -11Q69 -11 59 -2T48 16Q48 30 121 320L195 616Q195 629 188 632T149 637H128Q122 643 122 645T124 664Q129 683 137 683Z"
                  data-c="210E"
                />
              </g>
              <g
                data-mml-node="mi"
                transform="translate(576,0)"
              >
                <path
                  d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"
                  data-c="1D45F"
                />
              </g>
              <g
                data-mml-node="mi"
                transform="translate(1027,0)"
              >
                <path
                  d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"
                  data-c="1D452"
                />
              </g>
              <g
                data-mml-node="mi"
                transform="translate(1493,0)"
              >
                <path
                  d="M118 -162Q120 -162 124 -164T135 -167T147 -168Q160 -168 171 -155T187 -126Q197 -99 221 27T267 267T289 382V385H242Q195 385 192 387Q188 390 188 397L195 425Q197 430 203 430T250 431Q298 431 298 432Q298 434 307 482T319 540Q356 705 465 705Q502 703 526 683T550 630Q550 594 529 578T487 561Q443 561 443 603Q443 622 454 636T478 657L487 662Q471 668 457 668Q445 668 434 658T419 630Q412 601 403 552T387 469T380 433Q380 431 435 431Q480 431 487 430T498 424Q499 420 496 407T491 391Q489 386 482 386T428 385H372L349 263Q301 15 282 -47Q255 -132 212 -173Q175 -205 139 -205Q107 -205 81 -186T55 -132Q55 -95 76 -78T118 -61Q162 -61 162 -103Q162 -122 151 -136T127 -157L118 -162Z"
                  data-c="1D453"
                />
              </g>
              <g
                data-mjx-texclass="ORD"
                data-mml-node="TeXAtom"
                transform="translate(2043,0)"
              >
                <g
                  data-mml-node="mi"
                >
                  <path
                    d="M297 596Q297 627 318 644T361 661Q378 661 389 651T403 623Q403 595 384 576T340 557Q322 557 310 567T297 596ZM288 376Q288 405 262 405Q240 405 220 393T185 362T161 325T144 293L137 279Q135 278 121 278H107Q101 284 101 286T105 299Q126 348 164 391T252 441Q253 441 260 441T272 442Q296 441 316 432Q341 418 354 401T367 348V332L318 133Q267 -67 264 -75Q246 -125 194 -164T75 -204Q25 -204 7 -183T-12 -137Q-12 -110 7 -91T53 -71Q70 -71 82 -81T95 -112Q95 -148 63 -167Q69 -168 77 -168Q111 -168 139 -140T182 -74L193 -32Q204 11 219 72T251 197T278 308T289 365Q289 372 288 376Z"
                    data-c="1D457"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(412,0)"
                >
                  <path
                    d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"
                    data-c="1D44E"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(941,0)"
                >
                  <path
                    d="M173 380Q173 405 154 405Q130 405 104 376T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Q21 294 29 316T53 368T97 419T160 441Q202 441 225 417T249 361Q249 344 246 335Q246 329 231 291T200 202T182 113Q182 86 187 69Q200 26 250 26Q287 26 319 60T369 139T398 222T409 277Q409 300 401 317T383 343T365 361T357 383Q357 405 376 424T417 443Q436 443 451 425T467 367Q467 340 455 284T418 159T347 40T241 -11Q177 -11 139 22Q102 54 102 117Q102 148 110 181T151 298Q173 362 173 380Z"
                    data-c="1D463"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(1426,0)"
                >
                  <path
                    d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"
                    data-c="1D44E"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(1955,0)"
                >
                  <path
                    d="M131 289Q131 321 147 354T203 415T300 442Q362 442 390 415T419 355Q419 323 402 308T364 292Q351 292 340 300T328 326Q328 342 337 354T354 372T367 378Q368 378 368 379Q368 382 361 388T336 399T297 405Q249 405 227 379T204 326Q204 301 223 291T278 274T330 259Q396 230 396 163Q396 135 385 107T352 51T289 7T195 -10Q118 -10 86 19T53 87Q53 126 74 143T118 160Q133 160 146 151T160 120Q160 94 142 76T111 58Q109 57 108 57T107 55Q108 52 115 47T146 34T201 27Q237 27 263 38T301 66T318 97T323 122Q323 150 302 164T254 181T195 196T148 231Q131 256 131 289Z"
                    data-c="1D460"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(2424,0)"
                >
                  <path
                    d="M34 159Q34 268 120 355T306 442Q362 442 394 418T427 355Q427 326 408 306T360 285Q341 285 330 295T319 325T330 359T352 380T366 386H367Q367 388 361 392T340 400T306 404Q276 404 249 390Q228 381 206 359Q162 315 142 235T121 119Q121 73 147 50Q169 26 205 26H209Q321 26 394 111Q403 121 406 121Q410 121 419 112T429 98T420 83T391 55T346 25T282 0T202 -11Q127 -11 81 37T34 159Z"
                    data-c="1D450"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(2857,0)"
                >
                  <path
                    d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"
                    data-c="1D45F"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(3308,0)"
                >
                  <path
                    d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"
                    data-c="1D456"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(3653,0)"
                >
                  <path
                    d="M23 287Q24 290 25 295T30 317T40 348T55 381T75 411T101 433T134 442Q209 442 230 378L240 387Q302 442 358 442Q423 442 460 395T497 281Q497 173 421 82T249 -10Q227 -10 210 -4Q199 1 187 11T168 28L161 36Q160 35 139 -51T118 -138Q118 -144 126 -145T163 -148H188Q194 -155 194 -157T191 -175Q188 -187 185 -190T172 -194Q170 -194 161 -194T127 -193T65 -192Q-5 -192 -24 -194H-32Q-39 -187 -39 -183Q-37 -156 -26 -148H-6Q28 -147 33 -136Q36 -130 94 103T155 350Q156 355 156 364Q156 405 131 405Q109 405 94 377T71 316T59 280Q57 278 43 278H29Q23 284 23 287ZM178 102Q200 26 252 26Q282 26 310 49T356 107Q374 141 392 215T411 325V331Q411 405 350 405Q339 405 328 402T306 393T286 380T269 365T254 350T243 336T235 326L232 322Q232 321 229 308T218 264T204 212Q178 106 178 102Z"
                    data-c="1D45D"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(4156,0)"
                >
                  <path
                    d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z"
                    data-c="1D461"
                  />
                </g>
                <g
                  data-mml-node="mo"
                  transform="translate(4794.8,0)"
                >
                  <path
                    d="M78 370Q78 394 95 412T138 430Q162 430 180 414T199 371Q199 346 182 328T139 310T96 327T78 370ZM78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z"
                    data-c="3A"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(5350.6,0)"
                >
                  <path
                    d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"
                    data-c="1D44E"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(5879.6,0)"
                >
                  <path
                    d="M117 59Q117 26 142 26Q179 26 205 131Q211 151 215 152Q217 153 225 153H229Q238 153 241 153T246 151T248 144Q247 138 245 128T234 90T214 43T183 6T137 -11Q101 -11 70 11T38 85Q38 97 39 102L104 360Q167 615 167 623Q167 626 166 628T162 632T157 634T149 635T141 636T132 637T122 637Q112 637 109 637T101 638T95 641T94 647Q94 649 96 661Q101 680 107 682T179 688Q194 689 213 690T243 693T254 694Q266 694 266 686Q266 675 193 386T118 83Q118 81 118 75T117 65V59Z"
                    data-c="1D459"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(6177.6,0)"
                >
                  <path
                    d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"
                    data-c="1D452"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(6643.6,0)"
                >
                  <path
                    d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"
                    data-c="1D45F"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(7094.6,0)"
                >
                  <path
                    d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z"
                    data-c="1D461"
                  />
                </g>
                <g
                  data-mml-node="mo"
                  transform="translate(7455.6,0)"
                >
                  <path
                    d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"
                    data-c="28"
                  />
                </g>
                <g
                  data-mml-node="mn"
                  transform="translate(7844.6,0)"
                >
                  <path
                    d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"
                    data-c="31"
                  />
                </g>
                <g
                  data-mml-node="mo"
                  transform="translate(8344.6,0)"
                >
                  <path
                    d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"
                    data-c="29"
                  />
                </g>
              </g>
              <g
                data-mjx-texclass="ORD"
                data-mml-node="TeXAtom"
                transform="translate(10776.6,0)"
              >
                <g
                  data-mml-node="mi"
                >
                  <path
                    d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"
                    data-c="1D436"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(760,0)"
                >
                  <path
                    d="M117 59Q117 26 142 26Q179 26 205 131Q211 151 215 152Q217 153 225 153H229Q238 153 241 153T246 151T248 144Q247 138 245 128T234 90T214 43T183 6T137 -11Q101 -11 70 11T38 85Q38 97 39 102L104 360Q167 615 167 623Q167 626 166 628T162 632T157 634T149 635T141 636T132 637T122 637Q112 637 109 637T101 638T95 641T94 647Q94 649 96 661Q101 680 107 682T179 688Q194 689 213 690T243 693T254 694Q266 694 266 686Q266 675 193 386T118 83Q118 81 118 75T117 65V59Z"
                    data-c="1D459"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(1058,0)"
                >
                  <path
                    d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"
                    data-c="1D456"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(1403,0)"
                >
                  <path
                    d="M34 159Q34 268 120 355T306 442Q362 442 394 418T427 355Q427 326 408 306T360 285Q341 285 330 295T319 325T330 359T352 380T366 386H367Q367 388 361 392T340 400T306 404Q276 404 249 390Q228 381 206 359Q162 315 142 235T121 119Q121 73 147 50Q169 26 205 26H209Q321 26 394 111Q403 121 406 121Q410 121 419 112T429 98T420 83T391 55T346 25T282 0T202 -11Q127 -11 81 37T34 159Z"
                    data-c="1D450"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(1836,0)"
                >
                  <path
                    d="M121 647Q121 657 125 670T137 683Q138 683 209 688T282 694Q294 694 294 686Q294 679 244 477Q194 279 194 272Q213 282 223 291Q247 309 292 354T362 415Q402 442 438 442Q468 442 485 423T503 369Q503 344 496 327T477 302T456 291T438 288Q418 288 406 299T394 328Q394 353 410 369T442 390L458 393Q446 405 434 405H430Q398 402 367 380T294 316T228 255Q230 254 243 252T267 246T293 238T320 224T342 206T359 180T365 147Q365 130 360 106T354 66Q354 26 381 26Q429 26 459 145Q461 153 479 153H483Q499 153 499 144Q499 139 496 130Q455 -11 378 -11Q333 -11 305 15T277 90Q277 108 280 121T283 145Q283 167 269 183T234 206T200 217T182 220H180Q168 178 159 139T145 81T136 44T129 20T122 7T111 -2Q98 -11 83 -11Q66 -11 57 -1T48 16Q48 26 85 176T158 471L195 616Q196 629 188 632T149 637H144Q134 637 131 637T124 640T121 647Z"
                    data-c="1D458"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(2357,0)"
                >
                  <path
                    d="M289 629Q289 635 232 637Q208 637 201 638T194 648Q194 649 196 659Q197 662 198 666T199 671T201 676T203 679T207 681T212 683T220 683T232 684Q238 684 262 684T307 683Q386 683 398 683T414 678Q415 674 451 396L487 117L510 154Q534 190 574 254T662 394Q837 673 839 675Q840 676 842 678T846 681L852 683H948Q965 683 988 683T1017 684Q1051 684 1051 673Q1051 668 1048 656T1045 643Q1041 637 1008 637Q968 636 957 634T939 623Q936 618 867 340T797 59Q797 55 798 54T805 50T822 48T855 46H886Q892 37 892 35Q892 19 885 5Q880 0 869 0Q864 0 828 1T736 2Q675 2 644 2T609 1Q592 1 592 11Q592 13 594 25Q598 41 602 43T625 46Q652 46 685 49Q699 52 704 61Q706 65 742 207T813 490T848 631L654 322Q458 10 453 5Q451 4 449 3Q444 0 433 0Q418 0 415 7Q413 11 374 317L335 624L267 354Q200 88 200 79Q206 46 272 46H282Q288 41 289 37T286 19Q282 3 278 1Q274 0 267 0Q265 0 255 0T221 1T157 2Q127 2 95 1T58 0Q43 0 39 2T35 11Q35 13 38 25T43 40Q45 46 65 46Q135 46 154 86Q158 92 223 354T289 629Z"
                    data-c="1D440"
                  />
                </g>
                <g
                  data-mml-node="mi"
                  transform="translate(3408,0)"
                >
                  <path
                    d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"
                    data-c="1D452"
                  />
                </g>
                <g
                  data-mml-node="mo"
                  transform="translate(3874,0)"
                >
                  <path
                    d="M78 661Q78 682 96 699T138 716T180 700T199 661Q199 654 179 432T158 206Q156 198 139 198Q121 198 119 206Q118 209 98 431T78 661ZM79 61Q79 89 97 105T141 121Q164 119 181 104T198 61Q198 31 181 16T139 1Q114 1 97 16T79 61Z"
                    data-c="21"
                  />
                </g>
              </g>
            </g>
          </g>
        </svg>
      </span>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;
